### PR TITLE
add TextMatchFilterOptimizer to maximally push down `text_match` filters to Lucene

### PR DESCRIPTION
**Pull request by itschrispeck**
**Motivation:** 
Query performance against the Lucene index suffers when chaining multiple `text_match` predicates together. Our users often programmatically generate their queries, which exacerbates the issue as 10s/100s of `text_match` predicates can be included in a single query. 

Because of this, users are required to understand Pinot's Lucene implementation details for them to compose an efficient query. To remove this requirement, this PR adds a `TextMatchFilterOptmizer` that performs the optimization automatically. 

**Summary:**
This functionality is best understood through the unit testcases. In short: 
- Merge all AND's and OR's `text_match` operands when possible, without affecting query accuracy
- Push down NOT into Lucene, unless all `text_match` filters are inversed, then the NOT expression remains in Pinot

**Open question:**
There is one edge case (that I can think of) where this optimization can hurt performance: if there are a number of `text_match OR text_match OR text_match` etc, early termination when `limit` is reached might take longer since the entire merged `text_match` query must now complete. For this reason, it might be prudent to put this behind a query option (or rather a query option to disable it, since I believe it makes more sense to enable by default). 

Ideally, the `LuceneDocIdCollector` could early terminate (but doesn't currently have the required context). 

Testing: unit tests (query performance separately verified via running the optimized vs unoptimized queries). Sample optimization improvements:
```
count(*) = 500_000_000
optimized: 2950ms
unoptmized: 8350ms
> select count(*) from table where text_match(column, '"message_logtype:storage-event" AND /.*c28487d062.*/')
> select count(*) from table where text_match(column, '"message_logtype:storage-event"') AND text_match(column, '/.*c28487d062.*/')


count(*) = 2
optimized: 12ms
unoptimized: 8400ms
> select count(*) from table where text_match(column, '"message_logtype:storage-event" AND /.*c28487d062.*/ AND "offset:46127612"')
> select count(*) from table where text_match(column, '"message_logtype:storage-event"') AND text_match(column, '/.*c28487d062.*/') AND text_match(column, '"offset:46127612"')
```

tags: `feature`, `performance` (?)

### CODE REVIEW

**Comment by itschrispeck** [comment]
@siddharthteotia I'd appreciate your thoughts/review

File: pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/QueryOptimizerTest.java [hunk 1]
```diff
@@ -250,6 +265,37 @@ public void testQueries() {
...
+    testQuery("SELECT * FROM testTable WHERE int = 1 AND TEXT_MATCH(string, 'foo') AND TEXT_MATCH(string, 'bar')",
+        "SELECT * FROM testTable WHERE int = 1 AND TEXT_MATCH(string, 'foo AND bar')");
+    testQuery("SELECT * FROM testTable WHERE int = 1 OR TEXT_MATCH(string, 'foo') AND TEXT_MATCH(string, 'bar')",
+        "SELECT * FROM testTable WHERE int = 1 OR TEXT_MATCH(string, 'foo AND bar')");
+    testQuery("SELECT * FROM testTable WHERE TEXT_MATCH(string, 'foo') AND NOT TEXT_MATCH(string, 'bar')",
+        "SELECT * FROM testTable WHERE TEXT_MATCH(string, 'foo AND NOT bar')");
+    testQuery("SELECT * FROM testTable WHERE NOT TEXT_MATCH(string, 'foo') AND TEXT_MATCH(string, 'bar')",
+        "SELECT * FROM testTable WHERE TEXT_MATCH(string, 'NOT foo AND bar')");
+    testQuery("SELECT * FROM testTable WHERE NOT TEXT_MATCH(string, 'foo') AND NOT TEXT_MATCH(string, 'bar')",
+        "SELECT * FROM testTable WHERE NOT TEXT_MATCH(string, 'foo AND bar')");
```
Comment (chenboat) [inline pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/QueryOptimizerTest.java:285]: Change foo AND bar to foo OR bar?

File: pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/QueryOptimizerTest.java [hunk 2]
```diff
@@ -250,6 +265,37 @@ public void testQueries() {
...
+    testQuery("SELECT * FROM testTable WHERE TEXT_MATCH(string1, 'foo1') AND TEXT_MATCH(string1, 'bar1') OR "
+            + "TEXT_MATCH(string1, 'baz1') AND TEXT_MATCH(string2, 'foo')",
+        "SELECT * FROM testTable WHERE TEXT_MATCH(string1, 'foo1 AND bar1') OR TEXT_MATCH(string1, 'baz1') AND "
+            + "TEXT_MATCH(string2, 'foo')");
+    testQuery("SELECT * FROM testTable WHERE TEXT_MATCH(string1, 'foo1') AND TEXT_MATCH(string1, 'bar1')"
+            + "AND TEXT_MATCH(string2, 'foo2') AND TEXT_MATCH(string2, 'bar2')",
+        "SELECT * FROM testTable WHERE TEXT_MATCH(string1, 'foo1 AND bar1') AND TEXT_MATCH(string2, 'foo2 AND bar2')");
+    testCannotOptimizeQuery("SELECT * FROM testTable WHERE TEXT_MATCH(string1, 'foo') OR TEXT_MATCH(string2, 'bar')");
+    testCannotOptimizeQuery(
+        "SELECT * FROM testTable WHERE int = 1 AND TEXT_MATCH(string, 'foo') OR TEXT_MATCH(string, 'bar')");
```
Comment (chenboat) [inline pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/QueryOptimizerTest.java:297]: why this one can not be optimized? the columns are the same "string"?

File: pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/QueryOptimizerTest.java (hunk 1 above)
Comment (itschrispeck) [reply pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/QueryOptimizerTest.java:287]: oops, good catch!

File: pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/QueryOptimizerTest.java (hunk 2 above)
Comment (itschrispeck) [reply pinot-core/src/test/java/org/apache/pinot/core/query/optimizer/QueryOptimizerTest.java:299]: `int = 1 AND text_match(x) OR text_match(y)` wouldn't be equivalent to `int = 1 AND text_match(x OR y)`

**Comment by deemoliu** [comment]
this looks awesome. is it possible to optimize regexp_like(x) OR regexp_like(y) ?

**Pull request merged**